File: core/compilers/openapi.py
#!/usr/bin/env python3
"""
LAP Compiler — OpenAPI → LAP format

Usage:
    python compiler.py <openapi-spec.yaml> [-o output.lap]
"""

import argparse
import datetime
import json
import sys
from pathlib import Path

import yaml

from core.formats.lap import LAPSpec, Endpoint, Param, ResponseSchema, ResponseField, ErrorSchema


def resolve_ref(spec: dict, ref: str, _visited: set = None) -> dict:
    """Resolve a $ref pointer in an OpenAPI spec with cycle detection."""
    if _visited is None:
        _visited = set()
    if ref in _visited:
        raise ValueError(f"Circular $ref detected: {ref}")
    _visited.add(ref)
    parts = ref.lstrip("#/").split("/")
    node = spec
    for part in parts:
        node = node.get(part, {})
    # If the resolved node itself contains a $ref, resolve it too
    if isinstance(node, dict) and "$ref" in node:
        return resolve_ref(spec, node["$ref"], _visited)
    return node


def extract_type(schema: dict, spec: dict) -> str:
    """Extract a concise type string from an OpenAPI schema."""
    if "$ref" in schema:
        schema = resolve_ref(spec, schema["$ref"])

    t = schema.get("type", "any")
    # OpenAPI 3.1 allows type as a list, e.g. ["integer", "null"]
    if isinstance(t, list):
        non_null = [x for x in t if x != "null"]
        t = non_null[0] if non_null else "any"
    fmt = schema.get("format", "")

    if t == "string" and fmt:
        return f"str({fmt})"
    if t == "string":
        return "str"
    if t == "integer":
        return f"int({fmt})" if fmt else "int"
    if t == "number":
        return f"num({fmt})" if fmt else "num"
    if t == "boolean":
        return "bool"
    if t == "array":
        items_type = extract_type(schema.get("items", {}), spec)
        return f"[{items_type}]"
    if t == "object":
        return "map"
    return t


def extract_type_inline(schema: dict, spec: dict, depth: int = 0, max_depth: int = 1) -> str:
    """Extract type string, inlining object properties up to max_depth.

    Never returns bare 'map' for objects with known properties.
    Uses '!' suffix on required nested field names.
    """
    if "$ref" in schema:
        schema = resolve_ref(spec, schema["$ref"])

    t = schema.get("type", "any")
    if isinstance(t, list):
        non_null = [x for x in t if x != "null"]
        t = non_null[0] if non_null else "any"

    if t == "object" and schema.get("properties") and depth < max_depth:
        props = schema.get("properties", {})
        required_names = set(schema.get("required", []))
        parts = []
        for prop_name, prop_schema in props.items():
            if "$ref" in prop_schema:
                prop_schema = resolve_ref(spec, prop_schema["$ref"])
            prop_type = extract_type_inline(prop_schema, spec, depth + 1, max_depth)
            req_marker = "!" if prop_name in required_names else ""
            parts.append(f"{prop_name}{req_marker}: {prop_type}")
        return f"map{{{', '.join(parts)}}}"

    if t == "array":
        items_schema = schema.get("items", {})
        items_type = extract_type_inline(items_schema, spec, depth, max_depth)
        return f"[{items_type}]"

    # Fall back to regular extract_type for scalars or objects at max depth
    return extract_type(schema, spec)


def extract_params(param_list: list, spec: dict) -> tuple[list, list]:
    """Extract required and optional params from OpenAPI parameters."""
    required = []
    optional = []

    for p in param_list:
        if "$ref" in p:
            p = resolve_ref(spec, p["$ref"])

        if "name" not in p or not p["name"].strip():
            continue  # skip malformed params

        schema = p.get("schema", {})
        enum_vals = [v for v in schema.get("enum", []) if v is not None]
        param = Param(
            name=p["name"],
            type=extract_type(schema, spec),
            required=p.get("required", False),
            description=p.get("description", "").replace('\n', ' ').strip(),
            enum=enum_vals,
            default=str(schema["default"]) if "default" in schema else None,
        )

        if param.required:
            required.append(param)
        else:
            optional.append(param)

    return required, optional


def extract_request_body(body: dict, spec: dict) -> list:
    """Extract params from request body, inlining nested object schemas."""
    if not body:
        return []

    if "$ref" in body:
        body = resolve_ref(spec, body["$ref"])

    content = body.get("content", {})
    json_schema = content.get("application/json", {}).get("schema", {})

    if "$ref" in json_schema:
        json_schema = resolve_ref(spec, json_schema["$ref"])

    params = []
    required_names = set(json_schema.get("required", []))
    properties = json_schema.get("properties", {})

    for name, schema in properties.items():
        if not name.strip():
            continue
        if "$ref" in schema:
            schema = resolve_ref(spec, schema["$ref"])

        # Use inline expansion for object-type params to avoid bare 'map'
        type_str = extract_type_inline(schema, spec, depth=0, max_depth=1)

        enum_vals = [v for v in schema.get("enum", []) if v is not None]
        params.append(Param(
            name=name,
            type=type_str,
            required=name in required_names,
            description=schema.get("description", "").replace('\n', ' ').strip(),
            enum=enum_vals,
            default=str(schema["default"]) if "default" in schema else None,
        ))

    return params


def extract_response_fields(schema: dict, spec: dict, depth: int = 0, max_depth: int = 2, max_properties: int = 500) -> list:
    """Recursively extract typed fields from a response schema."""
    if "$ref" in schema:
        schema = resolve_ref(spec, schema["$ref"])

    fields = []
    properties = schema.get("properties", {})

    for count, (name, prop) in enumerate(properties.items()):
        if count >= max_properties:
            break
        if "$ref" in prop:
            prop = resolve_ref(spec, prop["$ref"])

        type_str = extract_type(prop, spec)
        nullable = prop.get("nullable", False) or "null" in str(prop.get("type", ""))

        children = []
        if prop.get("type") == "object" and depth < max_depth and prop.get("properties"):
            children = extract_response_fields(prop, spec, depth + 1, max_depth)

        fields.append(ResponseField(
            name=name,
            type=type_str,
            nullable=nullable,
            children=children,
        ))

    return fields


def extract_response_schemas(responses: dict, spec: dict) -> tuple:
    """Extract typed response and error schemas from OpenAPI responses."""
    response_schemas = []
    error_schemas = []

    for code, resp in responses.items():
        if "$ref" in resp:
            resp = resolve_ref(spec, resp["$ref"])

        desc = resp.get("description", "").replace('\n', ' ').strip()

        # Extract response body schema
        content = resp.get("content", {})
        json_content = content.get("application/json", {})
        schema = json_content.get("schema", {})

        if "$ref" in schema:
            schema = resolve_ref(spec, schema["$ref"])

        if code.startswith("2"):
            fields = extract_response_fields(schema, spec) if schema.get("properties") else []
            response_schemas.append(ResponseSchema(
                status_code=code,
                description=desc,
                fields=fields,
            ))
        elif code != "default":
            error_schemas.append(ErrorSchema(
                code=code,
                description=desc,
            ))

    return response_schemas, error_schemas


def extract_auth(spec: dict) -> str:
    """Extract auth scheme from security definitions."""
    schemes = spec.get("components", {}).get("securitySchemes", {})
    security = spec.get("security", [])

    if not schemes:
        return ""

    parts = []
    for scheme_name, scheme in schemes.items():
        t = scheme.get("type", "")
        if t == "http":
            parts.append(f"Bearer {scheme.get('scheme', 'token')}")
        elif t == "apiKey":
            parts.append(f"ApiKey {scheme.get('name', 'key')} in {scheme.get('in', 'header')}")
        elif t == "oauth2":
            parts.append("OAuth2")
        else:
            parts.append(t)

    return " | ".join(parts)


def _json_default(obj):
    """Handle non-serializable types in OpenAPI examples (datetime, date, etc.)."""
    if isinstance(obj, (datetime.datetime, datetime.date)):
        return obj.isoformat()
    return str(obj)


def extract_request_example(body: dict, spec: dict) -> str:
    """Extract a request example from OpenAPI requestBody as compact JSON."""
    if not body:
        return ""

    if "$ref" in body:
        body = resolve_ref(spec, body["$ref"])

    content = body.get("content", {})
    json_content = content.get("application/json", {})

    # Direct example on media type
    example = json_content.get("example")
    if example:
        return json.dumps(example, separators=(',', ':'), default=_json_default)

    # Named examples
    examples = json_content.get("examples", {})
    if examples:
        first = next(iter(examples.values()), {})
        if isinstance(first, dict):
            if "$ref" in first:
                first = resolve_ref(spec, first["$ref"])
            value = first.get("value")
            if value:
                return json.dumps(value, separators=(',', ':'), default=_json_default)

    return ""


# Threshold: a body param must appear in this fraction of body-having
# endpoints to be considered "common" and extracted into @common_fields.
_COMMON_FIELD_THRESHOLD = 0.95


def compile_openapi(spec_path: str) -> LAPSpec:
    """Compile an OpenAPI spec to LAP format."""
    path = Path(spec_path)
    file_size = path.stat().st_size
    if file_size > 50 * 1024 * 1024:
        raise ValueError(f"OpenAPI spec too large: {file_size} bytes (max 50MB)")
    raw = path.read_text()

    if path.suffix in (".yaml", ".yml"):
        spec = yaml.safe_load(raw)
    else:
        spec = json.loads(raw)

    if not isinstance(spec, dict):
        raise ValueError("Invalid OpenAPI spec: expected a YAML mapping")

    info = spec.get("info", {})
    servers = spec.get("servers", [])
    base_url = servers[0]["url"] if servers else ""

    lap = LAPSpec(
        api_name=info.get("title", path.stem),
        base_url=base_url,
        version=info.get("version", ""),
        auth_scheme=extract_auth(spec),
    )

    paths = spec.get("paths", {})
    for path_str, methods in paths.items():
        for method, details in methods.items():
            if method in ("get", "post", "put", "patch", "delete", "head", "options"):
                req_params, opt_params = extract_params(
                    details.get("parameters", []), spec
                )
                body_params = extract_request_body(
                    details.get("requestBody", {}), spec
                )

                response_schemas, error_schemas = extract_response_schemas(
                    details.get("responses", {}), spec
                )

                # Extract request example
                example = extract_request_example(
                    details.get("requestBody", {}), spec
                )

                endpoint = Endpoint(
                    method=method,
                    path=path_str,
                    summary=(details.get("summary") or details.get("description") or "").strip().split('\n')[0].strip(),
                    required_params=req_params,
                    optional_params=opt_params,
                    request_body=body_params,
                    response_schemas=response_schemas,
                    error_schemas=error_schemas,
                    example_request=example,
                )
                lap.endpoints.append(endpoint)

    # Deduplicate common fields -- any param (body, query, path, header)
    # appearing in >95% of all endpoints gets extracted into @common_fields.
    if len(lap.endpoints) > 5:
        from collections import Counter
        name_counts = Counter()
        for ep in lap.endpoints:
            seen = set()
            for p in ep.request_body + ep.required_params + ep.optional_params:
                if p.name not in seen:
                    name_counts[p.name] += 1
                    seen.add(p.name)

        threshold = len(lap.endpoints) * _COMMON_FIELD_THRESHOLD
        common_names = {name for name, count in name_counts.items()
                        if count >= threshold}

        if common_names:
            # Collect param objects from first endpoint that has each
            common_params = []
            found = set()
            for ep in lap.endpoints:
                for p in ep.request_body + ep.required_params + ep.optional_params:
                    if p.name in common_names and p.name not in found:
                        common_params.append(p)
                        found.add(p.name)
            # Strip from all endpoints
            for ep in lap.endpoints:
                ep.request_body = [p for p in ep.request_body
                                  if p.name not in common_names]
                ep.required_params = [p for p in ep.required_params
                                     if p.name not in common_names]
                ep.optional_params = [p for p in ep.optional_params
                                     if p.name not in common_names]
            lap.common_fields = common_params

    return lap


def main():
    parser = argparse.ArgumentParser(description="Compile OpenAPI spec to LAP format")
    parser.add_argument("spec", help="Path to OpenAPI spec (YAML/JSON)")
    parser.add_argument("-o", "--output", help="Output file (default: stdout)")
    parser.add_argument("--verbose", action="store_true", help="Also output verbose version for comparison")
    parser.add_argument("--lean", action="store_true", help="Strip all description comments for maximum compression")
    args = parser.parse_args()

    lap = compile_openapi(args.spec)
    result = lap.to_lap(lean=args.lean)

    if args.output:
        Path(args.output).write_text(result)
        print(f"✅ Compiled to {args.output}")
        if args.verbose:
            verbose_path = args.output.replace(".lap", ".verbose.md")
            Path(verbose_path).write_text(lap.to_original_text())
            print(f"📝 Verbose version: {verbose_path}")
    else:
        print(result)


if __name__ == "__main__":
    main()

File: sdks/typescript/src/compilers/openapi.ts
import * as fs from 'fs';
import * as path from 'path';
import * as yaml from 'js-yaml';
import {
  LAPSpec,
  Endpoint,
  Param,
  ResponseSchema,
  ResponseField,
  ErrorSchema,
} from '../parser';

function stripHtml(text: string): string {
  return text
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&amp;/g, '&')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/<[^>]+>/g, '');
}

// ── $ref resolution ──────────────────────────────────────────────────────────

function resolveRef(spec: Record<string, unknown>, ref: string, _visited?: Set<string>): Record<string, unknown> {
  if (_visited === undefined) _visited = new Set();
  if (_visited.has(ref)) throw new Error(`Circular $ref detected: ${ref}`);
  _visited.add(ref);

  const parts = ref.replace(/^#\//, '').split('/');
  let node: unknown = spec;
  for (const part of parts) {
    if (node === null || typeof node !== 'object') {
      node = {};
      break;
    }
    node = (node as Record<string, unknown>)[part] ?? {};
  }

  if (node !== null && typeof node === 'object' && '$ref' in (node as Record<string, unknown>)) {
    return resolveRef(spec, (node as Record<string, unknown>)['$ref'] as string, _visited);
  }
  return (node as Record<string, unknown>) ?? {};
}

// ── Type extraction ───────────────────────────────────────────────────────────

function extractType(schema: Record<string, unknown>, spec: Record<string, unknown>): string {
  if ('$ref' in schema) {
    schema = resolveRef(spec, schema['$ref'] as string);
  }

  let t: string = (schema['type'] as string) ?? 'any';

  // OpenAPI 3.1 allows type as an array, e.g. ["integer", "null"]
  if (Array.isArray(t)) {
    const nonNull = (t as string[]).filter((x) => x !== 'null');
    t = nonNull[0] ?? 'any';
  }

  const fmt = (schema['format'] as string) ?? '';

  if (t === 'string' && fmt) return `str(${fmt})`;
  if (t === 'string') return 'str';
  if (t === 'integer') return fmt ? `int(${fmt})` : 'int';
  if (t === 'number') return fmt ? `num(${fmt})` : 'num';
  if (t === 'boolean') return 'bool';
  if (t === 'array') {
    const items = (schema['items'] as Record<string, unknown>) ?? {};
    return `[${extractType(items, spec)}]`;
  }
  if (t === 'object') return 'map';
  return t;
}

function extractTypeInline(
  schema: Record<string, unknown>,
  spec: Record<string, unknown>,
  depth = 0,
  maxDepth = 1,
): string {
  if ('$ref' in schema) {
    schema = resolveRef(spec, schema['$ref'] as string);
  }

  let t: string = (schema['type'] as string) ?? 'any';
  if (Array.isArray(t)) {
    const nonNull = (t as string[]).filter((x) => x !== 'null');
    t = nonNull[0] ?? 'any';
  }

  if (t === 'object' && schema['properties'] && depth < maxDepth) {
    const props = (schema['properties'] as Record<string, unknown>) ?? {};
    const requiredNames = new Set<string>((schema['required'] as string[]) ?? []);
    const parts: string[] = [];

    for (const [propName, propSchemaRaw] of Object.entries(props)) {
      let propSchema = propSchemaRaw as Record<string, unknown>;
      if ('$ref' in propSchema) {
        propSchema = resolveRef(spec, propSchema['$ref'] as string);
      }
      const propType = extractTypeInline(propSchema, spec, depth + 1, maxDepth);
      const reqMarker = requiredNames.has(propName) ? '!' : '';
      parts.push(`${propName}${reqMarker}: ${propType}`);
    }
    return `map{${parts.join(', ')}}`;
  }

  if (t === 'array') {
    const itemsSchema = (schema['items'] as Record<string, unknown>) ?? {};
    const itemsType = extractTypeInline(itemsSchema, spec, depth, maxDepth);
    return `[${itemsType}]`;
  }

  // Fall back to scalar extraction for primitives or objects at max depth
  return extractType(schema, spec);
}

// ── Parameter extraction ──────────────────────────────────────────────────────

function extractParams(paramList: unknown[], spec: Record<string, unknown>): [Param[], Param[]] {
  const required: Param[] = [];
  const optional: Param[] = [];

  for (let p of paramList as Record<string, unknown>[]) {
    if ('$ref' in p) {
      p = resolveRef(spec, p['$ref'] as string);
    }

    const name = (p['name'] as string) ?? '';
    if (!name.trim()) continue; // skip malformed params

    const schema = (p['schema'] as Record<string, unknown>) ?? {};
    const rawEnum = (schema['enum'] as unknown[]) ?? [];
    const enumVals = rawEnum.filter((v) => v !== null).map(String);

    const isReq = Boolean(p['required']);
    const rawDefault = schema['default'];
    const defaultValue = rawDefault !== undefined ? String(rawDefault) : undefined;

    const typeStr = extractType(schema, spec);
    const isArray = typeStr.startsWith('[');

    const nullable = Boolean(schema['nullable']) ||
      (Array.isArray(schema['type']) && (schema['type'] as string[]).includes('null'));

    const param: Param = {
      name,
      type: typeStr,
      required: isReq,
      description: stripHtml(((p['description'] as string) ?? '')).replace(/\n/g, ' ').trim() || undefined,
      nullable,
      enumValues: enumVals.length > 0 ? enumVals : undefined,
      defaultValue,
      isArray,
    };

    if (isReq) {
      required.push(param);
    } else {
      optional.push(param);
    }
  }

  return [required, optional];
}

// ── Request body extraction ───────────────────────────────────────────────────

function extractRequestBody(body: Record<string, unknown>, spec: Record<string, unknown>): Param[] {
  if (!body || Object.keys(body).length === 0) return [];

  if ('$ref' in body) {
    body = resolveRef(spec, body['$ref'] as string);
  }

  const content = (body['content'] as Record<string, unknown>) ?? {};
  let jsonSchema = ((content['application/json'] as Record<string, unknown>) ?? {})['schema'] as
    | Record<string, unknown>
    | undefined;
  if (!jsonSchema) return [];

  if ('$ref' in jsonSchema) {
    jsonSchema = resolveRef(spec, jsonSchema['$ref'] as string);
  }

  const params: Param[] = [];
  const requiredNames = new Set<string>((jsonSchema['required'] as string[]) ?? []);
  const properties = (jsonSchema['properties'] as Record<string, unknown>) ?? {};

  for (const [name, schemaRaw] of Object.entries(properties)) {
    if (!name.trim()) continue;

    let schema = schemaRaw as Record<string, unknown>;
    if ('$ref' in schema) {
      schema = resolveRef(spec, schema['$ref'] as string);
    }

    // Use inline expansion for object-type params to avoid bare 'map'
    const typeStr = extractTypeInline(schema, spec, 0, 1);
    const isArray = typeStr.startsWith('[');

    const rawEnum = (schema['enum'] as unknown[]) ?? [];
    const enumVals = rawEnum.filter((v) => v !== null).map(String);

    const rawDefault = schema['default'];
    const defaultValue = rawDefault !== undefined ? String(rawDefault) : undefined;

    const isReq = requiredNames.has(name);

    const nullable = Boolean(schema['nullable']) ||
      (Array.isArray(schema['type']) && (schema['type'] as string[]).includes('null'));

    params.push({
      name,
      type: typeStr,
      required: isReq,
      description: stripHtml(((schema['description'] as string) ?? '')).replace(/\n/g, ' ').trim() || undefined,
      nullable,
      enumValues: enumVals.length > 0 ? enumVals : undefined,
      defaultValue,
      isArray,
    });
  }

  return params;
}

// ── Response field extraction ─────────────────────────────────────────────────

function extractResponseFields(
  schema: Record<string, unknown>,
  spec: Record<string, unknown>,
  depth = 0,
  maxDepth = 2,
  maxProperties = 500,
): ResponseField[] {
  if ('$ref' in schema) {
    schema = resolveRef(spec, schema['$ref'] as string);
  }

  const fields: ResponseField[] = [];
  const properties = (schema['properties'] as Record<string, unknown>) ?? {};
  let count = 0;

  for (const [name, propRaw] of Object.entries(properties)) {
    if (count >= maxProperties) break;
    count++;

    let prop = propRaw as Record<string, unknown>;
    if ('$ref' in prop) {
      prop = resolveRef(spec, prop['$ref'] as string);
    }

    const typeStr = extractType(prop, spec);
    const propTypeVal = prop['type'];
    const typeList = Array.isArray(propTypeVal) ? (propTypeVal as string[]) : [propTypeVal as string];
    const nullable = Boolean(prop['nullable']) || typeList.includes('null');

    let nested: ResponseField[] | undefined;
    if (prop['type'] === 'object' && depth < maxDepth && prop['properties']) {
      const children = extractResponseFields(prop, spec, depth + 1, maxDepth, maxProperties);
      if (children.length > 0) nested = children;
    }

    fields.push({
      name,
      type: typeStr,
      nullable,
      nested,
    });
  }

  return fields;
}

// ── Response schema extraction ────────────────────────────────────────────────

function extractResponseSchemas(
  responses: Record<string, unknown>,
  spec: Record<string, unknown>,
): [ResponseSchema[], ErrorSchema[]] {
  const responseSchemas: ResponseSchema[] = [];
  const errorSchemas: ErrorSchema[] = [];

  for (const [code, respRaw] of Object.entries(responses)) {
    let resp = respRaw as Record<string, unknown>;
    if ('$ref' in resp) {
      resp = resolveRef(spec, resp['$ref'] as string);
    }

    const desc = stripHtml(((resp['description'] as string) ?? '')).replace(/\n/g, ' ').trim() || undefined;

    const content = (resp['content'] as Record<string, unknown>) ?? {};
    const jsonContent = (content['application/json'] as Record<string, unknown>) ?? {};
    let schema = (jsonContent['schema'] as Record<string, unknown>) ?? {};

    if ('$ref' in schema) {
      schema = resolveRef(spec, schema['$ref'] as string);
    }

    if (code.startsWith('2')) {
      const fields = schema['properties'] ? extractResponseFields(schema, spec) : [];
      // Keep status code as string to support non-numeric codes like "2XX"
      const statusCode = code;
      responseSchemas.push({
        statusCode,
        description: desc,
        fields,
      });
    } else if (code !== 'default') {
      errorSchemas.push({
        statusCode: code,
        description: desc,
      });
    }
  }

  return [responseSchemas, errorSchemas];
}

// ── Auth extraction ───────────────────────────────────────────────────────────

function extractAuth(spec: Record<string, unknown>): string {
  const components = (spec['components'] as Record<string, unknown>) ?? {};
  const schemes = (components['securitySchemes'] as Record<string, unknown>) ?? {};

  if (Object.keys(schemes).length === 0) return '';

  const parts: string[] = [];
  for (const [_schemeName, schemeRaw] of Object.entries(schemes)) {
    const scheme = schemeRaw as Record<string, unknown>;
    const t = (scheme['type'] as string) ?? '';

    if (t === 'http') {
      parts.push(`Bearer ${(scheme['scheme'] as string) ?? 'token'}`);
    } else if (t === 'apiKey') {
      parts.push(`ApiKey ${(scheme['name'] as string) ?? 'key'} in ${(scheme['in'] as string) ?? 'header'}`);
    } else if (t === 'oauth2') {
      parts.push('OAuth2');
    } else if (t) {
      parts.push(t);
    }
  }

  return parts.join(' | ');
}

// ── Constants ─────────────────────────────────────────────────────────────────

const validMethods = new Set(['get', 'post', 'put', 'patch', 'delete', 'head', 'options']);
const COMMON_FIELD_THRESHOLD = 0.95;

// ── Request example extraction ────────────────────────────────────────────────

function extractRequestExample(
  body: Record<string, unknown>,
  spec: Record<string, unknown>,
  visited?: Set<string>,
): string | undefined {
  if (!body || Object.keys(body).length === 0) return undefined;

  if ('$ref' in body) {
    body = resolveRef(spec, body['$ref'] as string, visited);
  }

  const content = (body['content'] as Record<string, unknown>) ?? {};
  const jsonContent = (content['application/json'] as Record<string, unknown>) ?? {};

  const example = jsonContent['example'];
  if (example) {
    return JSON.stringify(example);
  }

  const examples = (jsonContent['examples'] as Record<string, unknown>) ?? {};
  for (const [, exVal] of Object.entries(examples)) {
    const ex = exVal as Record<string, unknown>;
    if (ex['$ref']) {
      const resolved = resolveRef(spec, ex['$ref'] as string, visited);
      if (resolved['value']) return JSON.stringify(resolved['value']);
    }
    if (ex['value']) return JSON.stringify(ex['value']);
  }
  return undefined;
}

// ── Main compiler ─────────────────────────────────────────────────────────────

export function compileOpenapi(specPath: string): LAPSpec {
  const filePath = path.resolve(specPath);
  const stat = fs.statSync(filePath);

  if (stat.size > 50 * 1024 * 1024) {
    throw new Error(`OpenAPI spec too large: ${stat.size} bytes (max 50MB)`);
  }

  const raw = fs.readFileSync(filePath, 'utf-8');
  const ext = path.extname(filePath).toLowerCase();

  let spec: Record<string, unknown>;

  if (ext === '.yaml' || ext === '.yml') {
    try {
      spec = yaml.load(raw) as Record<string, unknown>;
    } catch {
      // Strip unknown YAML tags and retry
      const cleaned = raw.replace(/![a-zA-Z_][\w-]*/g, '');
      spec = yaml.load(cleaned) as Record<string, unknown>;
    }
  } else {
    spec = JSON.parse(raw) as Record<string, unknown>;
  }

  if (!spec || typeof spec !== 'object' || Array.isArray(spec)) {
    throw new Error('Invalid OpenAPI spec: expected a YAML/JSON mapping');
  }

  const info = (spec['info'] as Record<string, unknown>) ?? {};
  const servers = (spec['servers'] as Record<string, unknown>[]) ?? [];
  const baseUrl = servers.length > 0 ? (servers[0]['url'] as string) ?? '' : '';
  const apiName = (info['title'] as string) ?? path.basename(filePath, ext);
  const apiVersion = (info['version'] as string) ?? undefined;
  const auth = extractAuth(spec) || undefined;

  const result: LAPSpec = {
    version: 'v0.3',
    apiName,
    baseUrl,
    apiVersion,
    auth,
    endpoints: [],
    getEndpoint(method: string, p: string): Endpoint | undefined {
      return this.endpoints.find((e) => e.method === method && e.path === p);
    },
  };

  const paths = (spec['paths'] as Record<string, unknown>) ?? {};

  for (const [pathStr, methodsRaw] of Object.entries(paths)) {
    const methods = methodsRaw as Record<string, unknown>;
    for (const [method, detailsRaw] of Object.entries(methods)) {
      if (!validMethods.has(method)) continue;

      const details = detailsRaw as Record<string, unknown>;
      const paramList = (details['parameters'] as unknown[]) ?? [];
      const [reqParams, optParams] = extractParams(paramList, spec);
      const requestBodyRaw = (details['requestBody'] as Record<string, unknown>) ?? {};
      const bodyParams = extractRequestBody(requestBodyRaw, spec);
      const [responseSchemas, errorSchemas] = extractResponseSchemas(
        (details['responses'] as Record<string, unknown>) ?? {},
        spec,
      );

      // Extract request example from requestBody
      const exampleRequest = extractRequestExample(requestBodyRaw, spec);

      const summary = stripHtml(
        (details['summary'] as string) ??
        (details['description'] as string) ??
        ''
      )
        .trim()
        .split('\n')[0]
        .trim();

      // Store body params separately for common-field deduplication
      const endpoint: Endpoint = {
        method,
        path: pathStr,
        description: summary || undefined,
        requiredParams: reqParams,
        optionalParams: optParams,
        allParams: [...reqParams, ...optParams, ...bodyParams],
        requestBody: bodyParams.length > 0 ? bodyParams : undefined,
        responses: responseSchemas,
        errors: errorSchemas,
        exampleRequest,
      };

      result.endpoints.push(endpoint);
    }
  }

  // ── Common field deduplication ──────────────────────────────────────────────
  // Any param (body, query, path, header) appearing in >95% of all endpoints
  // gets extracted into commonFields.
  if (result.endpoints.length > 5) {
    const nameCounts = new Map<string, number>();
    for (const ep of result.endpoints) {
      const seen = new Set<string>();
      const allParams = [
        ...(ep.requestBody || []),
        ...ep.requiredParams,
        ...ep.optionalParams,
      ];
      for (const p of allParams) {
        if (!seen.has(p.name)) {
          nameCounts.set(p.name, (nameCounts.get(p.name) || 0) + 1);
          seen.add(p.name);
        }
      }
    }

    const threshold = result.endpoints.length * COMMON_FIELD_THRESHOLD;
    const commonNames = new Set<string>();
    for (const [name, count] of nameCounts) {
      if (count >= threshold) commonNames.add(name);
    }

    if (commonNames.size > 0) {
      // Extract representative params for common fields
      const commonParams: Param[] = [];
      const addedNames = new Set<string>();
      for (const ep of result.endpoints) {
        const allParams = [
          ...(ep.requestBody || []),
          ...ep.requiredParams,
          ...ep.optionalParams,
        ];
        for (const p of allParams) {
          if (commonNames.has(p.name) && !addedNames.has(p.name)) {
            commonParams.push(p);
            addedNames.add(p.name);
          }
        }
      }

      // Strip common fields from individual endpoints
      for (const ep of result.endpoints) {
        ep.requiredParams = ep.requiredParams.filter(p => !commonNames.has(p.name));
        ep.optionalParams = ep.optionalParams.filter(p => !commonNames.has(p.name));
        if (ep.requestBody) {
          ep.requestBody = ep.requestBody.filter(p => !commonNames.has(p.name));
          if (ep.requestBody.length === 0) ep.requestBody = undefined;
        }
        // Rebuild allParams after stripping
        ep.allParams = [
          ...ep.requiredParams,
          ...ep.optionalParams,
          ...(ep.requestBody || []),
        ];
      }

      result.commonFields = commonParams;
    }
  }

  return result;
}
